aws-lc-rs/src/aead.rs

```diff
@@ -647,6 +678,317 @@ impl Debug for LessSafeKey {
+    /// `error::Unspecified` if encryption operation fails.
+    #[inline]
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn seal_in_place_separate_tag<A>(
+        &self,
+        aad: Aad<A>,
+        in_out: &mut [u8],
+    ) -> Result<(Nonce, Tag), Unspecified>
+    where
+        A: AsRef<[u8]>,
+    {
+        let nonce = if let AlgorithmID::CHACHA20_POLY1305 = self.algorithm.id {
+            Some(Nonce(FixedLength::<NONCE_LEN>::new()?))
+        } else {
+            None
+        };
+        seal_in_place_separate_tag(
+            self.algorithm,
+            &self.ctx,
+            nonce,
+            Aad::from(aad.as_ref()),
+            in_out,
+        )
+    }
+
+    /// The key's AEAD algorithm.
+    #[inline]
+    #[must_use]
+    pub fn algorithm(&self) -> &'static Algorithm {
+        self.algorithm
+    }
+}
+
+/// The Transport Layer Security (TLS) protocol version.
+pub enum TLSProtocolId {
+    /// TLS 1.2 (RFC 5246)
+    TLS12,
+
+    /// TLS 1.3 (RFC 8446)
+    TLS13,
+}
+
+/// AEAD Encryption key used for TLS porotocol record encryption.
+pub struct TLSRecordSealingKey {
+    ctx: Mutex<AeadCtx>,
+    algorithm: &'static Algorithm,
+}
+
+impl TLSRecordSealingKey {
+    /// New TLS record sealing key. Only supports `AES_128_GCM` and `AES_256_GCM`.
+    ///
+    /// # Errors
+    /// * `Unspecified`: Returned if the length of `key_bytes` does not match the chosen algorithm,
+    /// or if an unsupported algorithm is provided.
+    pub fn new(
+        algorithm: &'static Algorithm,
+        protocol: TLSProtocolId,
+        key_bytes: &[u8],
+    ) -> Result<Self, Unspecified> {
+        let ctx = Mutex::new(match (algorithm.id, protocol) {
+            (AlgorithmID::AES_128_GCM, TLSProtocolId::TLS12) => {
+                AeadCtx::aes_128_gcm_tls12(key_bytes)
+            }
+            (AlgorithmID::AES_128_GCM, TLSProtocolId::TLS13) => {
+                AeadCtx::aes_128_gcm_tls13(key_bytes)
+            }
+            (AlgorithmID::AES_256_GCM, TLSProtocolId::TLS12) => {
+                AeadCtx::aes_256_gcm_tls12(key_bytes)
+            }
+            (AlgorithmID::AES_256_GCM, TLSProtocolId::TLS13) => {
+                AeadCtx::aes_256_gcm_tls13(key_bytes)
+            }
+            (AlgorithmID::CHACHA20_POLY1305, _) => Err(Unspecified),
+        }?);
+        Ok(Self { ctx, algorithm })
+    }
+
+    /// Accepts a `Nonce` and `Aad` construction that is unique for this key and
+    /// TLS record sealing operation for the configured TLS protocol version.
+    ///
+    /// `nonce` must be unique and incremented per each sealing operation,
+    /// otherwise an error is returned.
+    ///
+    /// # Errors
+    /// `error::Unspecified` if encryption operation fails.
+    #[inline]
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn seal_in_place_append_tag<A, InOut>(
+        &self,
+        nonce: Nonce,
+        aad: Aad<A>,
+        in_out: &mut InOut,
+    ) -> Result<(), Unspecified>
```

Comment:
For TLS, isn't the AAD length expected to be 11 bytes?  Do we enforce that anywhere in our logic? Do we need a separate more-restrictive `Aad` type?

Reply:
I don't see AWS-LC enforcing this at the cipher usage? Nor do I see where you got 11 bytes from? In AWS-LC's AAD construction in libssl specifically it [appears](https://github.com/aws/aws-lc/blob/b84674b2d2512cfd200b7c392215b6eebda3c59a/ssl/ssl_aead_ctx.cc#L263-L264) to allocate 13 bytes for AAD as the upper bound?

But again this doesn't appear to be enforced at the cipher usage level.  It also seems to depend whether the AAD is expected to be just the header or if it includes other information (seems to depend on TLS 1.2 versus TLS 1.3).

Similar to nonce construction should we lave the proper aad construction left to the implementer? If we know what the proper max limit should be we could maybe sanity check it, but if it is variable I don't think we can enforce an exact size.

Reply:
Ok. This comment and the one above were based on my reading of documentation provided by BoringSSL here: https://commondatastorage.googleapis.com/chromium-boringssl-docs/aead.h.html
> These AEAD primitives do not meet the definition of generic AEADs. They are all specific to TLS and should not be used outside of that context. They must be initialized with [EVP_AEAD_CTX_init_with_direction](https://commondatastorage.googleapis.com/chromium-boringssl-docs/aead.h.html#EVP_AEAD_CTX_init_with_direction), are stateful, and may not be used concurrently. Any nonces are used as IVs, so they must be unpredictable. They only accept an ad parameter of length 11 (the standard TLS one with length omitted).

I'll look into it some more.